include/oneapi/dpl/pstl/algorithm_impl.h

```diff
@@ -134,43 +134,52 @@ __brick_walk1(_DifferenceType __n, _Function __f, ::std::true_type) noexcept
+    __internal::__brick_walk1(__first, __last, __f, typename _Tag::__is_vector{});
 }
 
-template <class _ExecutionPolicy, class _ForwardIterator, class _Function, class _IsVector>
-oneapi::dpl::__internal::__enable_if_host_execution_policy_conditional<
-    _ExecutionPolicy, __is_random_access_iterator<_ForwardIterator>::value, void>
-__pattern_walk1(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last, _Function __f,
-                _IsVector __is_vector,
-                /*parallel=*/::std::true_type)
+template <class _ExecutionPolicy, class _ForwardIterator, class _Function>
+void
+__pattern_walk1(__parallel_forward_tag<_ExecutionPolicy, _ForwardIterator> __tag, _ExecutionPolicy&& __exec,
+                _ForwardIterator __first, _ForwardIterator __last, _Function __f)
 {
+    typedef typename ::std::iterator_traits<_ForwardIterator>::reference _ReferenceType;
+    auto __func = [&__f](_ReferenceType arg) { __f(arg); };
     __internal::__except_handler([&]() {
-        __par_backend::__parallel_for(::std::forward<_ExecutionPolicy>(__exec), __first, __last,
-                                      [__f, __is_vector](_ForwardIterator __i, _ForwardIterator __j) {
-                                          __internal::__brick_walk1(__i, __j, __f, __is_vector);
-                                      });
+        __par_backend::__parallel_for_each(__tag, ::std::forward<_ExecutionPolicy>(__exec), __first, __last, __func);
     });
 }
 
-template <class _ExecutionPolicy, class _ForwardIterator, class _Function, class _IsVector>
-oneapi::dpl::__internal::__enable_if_host_execution_policy_conditional<
-    _ExecutionPolicy, !__is_random_access_iterator<_ForwardIterator>::value, void>
-__pattern_walk1(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last, _Function __f, _IsVector,
-                /*parallel=*/::std::true_type)
+template <class _ExecutionPolicy, class _ForwardIterator, class _Function>
+void
+__pattern_walk1(__parallel_tag<_ExecutionPolicy, _ForwardIterator> __tag, _ExecutionPolicy&& __exec,
+                _ForwardIterator __first, _ForwardIterator __last, _Function __f)
 {
-    typedef typename ::std::iterator_traits<_ForwardIterator>::reference _ReferenceType;
-    auto __func = [&__f](_ReferenceType arg) { __f(arg); };
+    using __tag_type = decltype(__tag);
     __internal::__except_handler([&]() {
-        __par_backend::__parallel_for_each(::std::forward<_ExecutionPolicy>(__exec), __first, __last, __func);
+        __par_backend::__parallel_for(__tag, ::std::forward<_ExecutionPolicy>(__exec), __first, __last,
+                                      [__f](_ForwardIterator __i, _ForwardIterator __j) {
+                                          __internal::__brick_walk1(__i, __j, __f, typename __tag_type::__is_vector{});
+                                      });
     });
 }
 
+template <class _Tag, class _ExecutionPolicy, class _ForwardIterator, class _UnaryPredicate, class _Tp>
```

Comment:
This pattern was added to show how it would look like. In general, we are not planning to add new patterns right away and would add those on demand.